Set Log Level API

ArtiomDivak opened the pull request:
Api feature for setting the log level for the nodes and the manager.
for more details please read https://github.com/containers/hirte/issues/189#issuecomment-1560632309

mwperina commented on src/agent/agent.c:
```diff
@@ -1333,6 +1333,47 @@ static int agent_method_disable_metrics(sd_bus_message *m, void *userdata, UNUSE
+//         hirte_log_infof("Log level changed to %s", r);
+//         return r;
+// }
+
```
Please remove commented out code lines from the patch

mwperina commented on src/agent/agent.c:
```diff
@@ -1348,6 +1389,8 @@ static const sd_bus_vtable internal_agent_vtable[] = {
         SD_BUS_METHOD("EnableMetrics", "", "", agent_method_enable_metrics, 0),
         SD_BUS_METHOD("DisableMetrics", "", "", agent_method_disable_metrics, 0),
+        // SD_BUS_METHOD("SetNodeLogLevel", "ss", "", set_node_log_level, 0),
+        // SD_BUS_METHOD("SetNodeLogLevel", "s", "", set_all_node_log_level, 0),
```
Please remove commented out code lines from the patch

mwperina commented on src/manager/node.c:
```diff
@@ -62,6 +62,7 @@ static const sd_bus_vtable node_vtable[] = {
         SD_BUS_METHOD("EnableUnitFiles", "asbb", "ba(sss)", node_method_passthrough_to_agent, 0),
         SD_BUS_METHOD("DisableUnitFiles", "asb", "a(sss)", node_method_passthrough_to_agent, 0),
         SD_BUS_METHOD("Reload", "", "", node_method_passthrough_to_agent, 0),
+        // SD_BUS_METHOD("SetNodeLogLevel", "ss", "", set_node_log_level, 0),
```
Please remove commented out code lines from the patch

mwperina commented on src/manager/node.c:
```diff
@@ -1511,6 +1512,44 @@ static int node_method_reload_unit(sd_bus_message *m, void *userdata, UNUSED sd_
+//         hirte_log_set_level(loglevel);
+//         return r;
+// }
+
```
Please remove commented out code lines from the patch

mwperina commented on src/manager/manager.c:
```diff
@@ -812,13 +811,35 @@ static int manager_method_metrics_disable(sd_bus_message *m, void *userdata, UNU
+                return 1;
+        }
+        LogLevel loglevel = string_to_log_level(level);
+        if (loglevel == LOG_LEVEL_INVALID) {
```
Shouldn't we emit some error message to the logs here?

mwperina reviewed (dismissed):
Please mention in the commit subject that this patch add log level API to the manager only. Also please fit the typo in the commit message

ArtiomDivak committed:
Set Log Level For The Manager

I added public API to the Manager.
Now a user  can use the with sdbus to set new log leve for the manger

Signed-off-by: Artiom Divak <adivak@redhat.com>

mwperina commented on src/manager/manager.c:
```diff
@@ -812,13 +811,35 @@ static int manager_method_metrics_disable(sd_bus_message *m, void *userdata, UNU
+        int r = sd_bus_message_read(m, "s", &level);
+        if (r < 0) {
+                return sd_bus_reply_method_errorf(
+                                m, SD_BUS_ERROR_FAILED, "Failed to read the parameter: %s", strerror(-r));
```
Wouldn't it be good to mention which parameter failed to be read?

ArtiomDivak committed:
Set Log Level For Node And Agent

The APIs of the node and the agent are connected.
The API of Agent is internal. So if the user wants
to change the log level for the agent he needs to
use the public API of the node

Signed-off-by: Artiom Divak <adivak@redhat.com>

mwperina commented on src/agent/agent.c:
```diff
@@ -1334,6 +1334,27 @@ static int agent_method_disable_metrics(sd_bus_message *m, void *userdata, UNUSE
+        int r = sd_bus_message_read(m, "s", &level);
+        if (r < 0) {
+                return sd_bus_reply_method_errorf(
+                                m, SD_BUS_ERROR_FAILED, "Failed to read the parameter: %s", strerror(-r));
```
Same as in manager, wouldn't it be good to mention which parameter failed to be read?

mwperina commented on src/agent/agent.c:
```diff
@@ -1334,6 +1334,27 @@ static int agent_method_disable_metrics(sd_bus_message *m, void *userdata, UNUSE
+        }
+        LogLevel loglevel = string_to_log_level(level);
+        if (loglevel == LOG_LEVEL_INVALID) {
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Invalid input for log level");
```
Same as in manager, shouldn't we emit also an error message to the logs?

ArtiomDivak committed:
Set LogLevel from hirtectl

hirtectl will support set-loglevel command. This command is to set log
level on specific node or on the manager. For more info and usage
please go to
https://github.com/containers/hirte/issues/189#issuecomment-1560632309.

Signed-off-by: Artiom Divak <adivak@redhat.com>

mkemel commented on data/org.containers.hirte.Node.xml:
```diff
@@ -58,7 +58,9 @@
       <arg name="units" type="a(ssssssouso)" direction="out" />
     </method>
-    <method name="Reload" />
+    <method name="SelNodeLogLevel">
```
```suggestion
    <method name="SetNodeLogLevel">
```

engelmi replied:
Since it is on the Node API, simply `SetLogLevel` is enough :)

mkemel replied:
Agree

mkemel commented on src/client/client.c:
```diff
@@ -614,6 +694,13 @@ int client_call_manager(Client *client) {
                 }
                 r = method_daemon_reload_on(client, client->opargv[0]);
+        } else if (streq(client->op, "set-loglevel")) {
+                if (client->opargc == 1) {
```
What about `if (client->opargc < 1)`?

engelmi replied:
Wouldn't that be the `else` case and return an `-EINVAL`?

mkemel replied:
That one wasn't there when I left this comment :)

mkemel replied:
Can be resolved now

mkemel commented on src/client/client.c:
```diff
@@ -614,6 +694,13 @@ int client_call_manager(Client *client) {
+                if (client->opargc == 1) {
+                        r = method_set_log_level(client, NULL, client->opargv[0]);
+                }
+                if (client->opargc > 1) {
```
I would use `if - else if - else if`

mkemel commented on src/client/client.c:
```diff
@@ -502,6 +502,86 @@ static int method_daemon_reload_on(Client *client, char *node_name) {
+                if (r < 0) {
+                        return r;
+                }
+        } else {
```
I think set-loglevel on all should be implemented on manager side, and there should be a separate method on the manager interface for that. Moving this logic to the client 1. Complicates the client 2. Makes it harder to do this via API (basically if you use the API and not the client, you should re-implement this logic yourself)

ArtiomDivak replied:
My first try to implement this was to call the manager via sdbus and then the manager will use sdbus to call the methods for the Nodes. I got some kind of symbolic link error. If you think it's better to write the implementation in the Manager I can try and try to do it again.

mkemel replied:
I definitely think that the client should be as thin as possible. Please do, and if you get errors I will be happy to assist with debug.
Look at `EnableMetrics`, it does something similar

engelmi replied:
I think this is going into the wrong direction. Please don't implement any "on all nodes" functionality, but only set log level for 
- the manager (hirte)
- a specific (hirte-agent)

as described in https://github.com/containers/hirte/issues/189#issuecomment-1560632309. Its ok to have the user call `hirtectl set-loglevel <node> <level>` multiple times for each node.

mkemel replied:
That's also an option. In any case, looping over existing nodes in client is wrong.

What's wrong with set-loglevel on all?

engelmi replied:
Nothing :) We just didn't discuss this option in our meeting so I didn't specify it in the issue - and to avoid changing the scope, I'd not put this in this PR. If we want to add it later, we can do so, of course.

mkemel replied:
I was not aware of the meeting, but client implementation caught my eye.
I agree with this approach - finish up this commit with functionality you described above, and then add the set-node-loglevel all functionality in a separate commit

engelmi commented on src/manager/manager.c:
```diff
@@ -812,13 +811,37 @@ static int manager_method_metrics_disable(sd_bus_message *m, void *userdata, UNU
+
+        int r = sd_bus_message_read(m, "s", &level);
+        if (r < 0) {
+                hirte_log_errorf("Failed to read the parameter: %s", strerror(-r));
+                return sd_bus_reply_method_errorf(
+                                m, SD_BUS_ERROR_FAILED, "Failed to read the parameter: %s", strerror(-r));
```
```suggestion
                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
```
@mwperina That would be gread, but as far as I know in case of multiple params we are not able to, unfortunately. Thats why we return just an `Invalid arguments` message. 
@ArtiomDivak The `hirte_log_errorf` isn't necessary here since its a "user related" error.

engelmi commented on src/manager/manager.c:
```diff
@@ -812,13 +811,37 @@ static int manager_method_metrics_disable(sd_bus_message *m, void *userdata, UNU
+        LogLevel loglevel = string_to_log_level(level);
+        if (loglevel == LOG_LEVEL_INVALID) {
+                hirte_log_errorf("Invalid input for log level: %s", loglevel);
+                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Invalid input for log level");
```
```suggestion
                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid input for log level");
```

engelmi commented on src/manager/node.c:
```diff
@@ -1511,6 +1513,50 @@ static int node_method_reload_unit(sd_bus_message *m, void *userdata, UNUSED sd_
+                if (r < 0) {
+                        hirte_log_errorf("Failed to send reply to proxy request: %s", strerror(-r));
+                }
+        }
```
```suggestion
        if (r < 0) {
                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
        }
```
Just as in the manager, just return with an `sd_bus_reply_method_errorf` stating invalid arguments.

engelmi commented on src/manager/manager.c:
```diff
@@ -812,13 +811,37 @@ static int manager_method_metrics_disable(sd_bus_message *m, void *userdata, UNU
+        }
+        LogLevel loglevel = string_to_log_level(level);
+        if (loglevel == LOG_LEVEL_INVALID) {
+                hirte_log_errorf("Invalid input for log level: %s", loglevel);
```
```suggestion
```
This log isn't needed since its an user error where he sent an invalid input.

engelmi commented on src/manager/node.c:
```diff
@@ -1511,6 +1513,50 @@ static int node_method_reload_unit(sd_bus_message *m, void *userdata, UNUSED sd_
+        }
+        LogLevel loglevel = string_to_log_level(level);
+        if (loglevel == LOG_LEVEL_INVALID) {
+                hirte_log_error("Failed to set log level call: Incorrect loglevel name");
```
```suggestion
```
This log isn't needed since its an user error where he sent an invalid input.

engelmi commented on src/manager/node.c:
```diff
@@ -1511,6 +1513,50 @@ static int node_method_reload_unit(sd_bus_message *m, void *userdata, UNUSED sd_
+                r = sd_bus_reply_method_return(m, "");
+                if (r < 0) {
+                        hirte_log_errorf("Failed to send reply to proxy request: %s", strerror(-r));
+                }
```
```suggestion
                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_INVALID_ARGS, "Invalid arguments");
```

engelmi commented on src/manager/node.c:
```diff
@@ -1511,6 +1513,50 @@ static int node_method_reload_unit(sd_bus_message *m, void *userdata, UNUSED sd_
+                        level);
+        if (r < 0) {
+                hirte_log_errorf("Failed to set log level call: %s", error.message);
+                sd_bus_error_free(&error);
```
```suggestion
```
You don't need to free the error, I think.

engelmi commented on src/manager/node.c:
```diff
@@ -1511,6 +1513,50 @@ static int node_method_reload_unit(sd_bus_message *m, void *userdata, UNUSED sd_
+        if (r < 0) {
+                hirte_log_errorf("Failed to set log level call: %s", error.message);
+                sd_bus_error_free(&error);
+                return false;
```
```suggestion
                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Internal error");
```
In case the call to the agent fails we need to forward an error to the caller. You can also add more detail to the message, e.g. instead of just `Internal error` you could state `Setting log level call to agent failed`.

engelmi commented on data/org.containers.hirte.Node.xml:
```diff
@@ -58,7 +58,9 @@
     <method name="ListUnits">
       <arg name="units" type="a(ssssssouso)" direction="out" />
     </method>
-    <method name="Reload" />
```
Please re-add the `Reload` method.

engelmi commented on src/manager/node.c:
```diff
@@ -61,7 +62,8 @@ static const sd_bus_vtable node_vtable[] = {
         SD_BUS_METHOD("DisableUnitFiles", "asb", "a(sss)", node_method_passthrough_to_agent, 0),
-        SD_BUS_METHOD("Reload", "", "", node_method_passthrough_to_agent, 0),
+        SD_BUS_METHOD("SetLogLevel", "s", "", node_method_set_log_level, 0),
+        SD_BUS_METHOD("Reload", "", "", NULL, 0),
```
```suggestion
        SD_BUS_METHOD("Reload", "", "", node_method_passthrough_to_agent, 0),
```
This seems wrong. Please re-add the method as it was before.

engelmi reviewed (commented):
LGTM
